core/invoker/src/main/scala/whisk/core/container/HttpUtils.scala

```diff
@@ -0,0 +1,135 @@
+
+/**
+ * This HTTP client is used only in the invoker to communicate with the action container.
+ * It allows to POST a JSON object and receive JSON object back; that is the
+ * content type and the accept headers are both 'application/json.
+ * The reason we still use this class for the action container is a mysterious hang
+ * in the Akka http client where a future fails to properly timeout and we have not
+ * determined why that is.
+ *
+ * @param hostname the host name
+ * @param timeoutMsec the timeout in msecs to wait for a response
+ */
+protected[container] class HttpUtils(
+    hostname: String,
+    timeout: FiniteDuration,
+    maxResponse: ByteSize) {
+
+    def close = Try(connection.close)
+
+    /**
+     * Posts to hostname/endpoint the given JSON object.
+     * Waits up to timeout before aborting on a good connection.
+     * If the endpoint is not ready, retry up to timeout.
+     * Every retry reduces the available timeout so that this method should not
+     * wait longer than the total timeout (within a small slack allowance).
+     *
+     * @param endpoint the path the api call relative to hostname
+     * @param body the json object to post
+     * @return Left(Error Message) or Right(Status Code, Response as UTF-8 String)
+     */
+    def post(endpoint: String, body: JsObject): Either[ContainerConnectionError, ContainerResponse] = {
+        val entity = new StringEntity(body.compactPrint, StandardCharsets.UTF_8)
+        entity.setContentType("application/json")
+
+        val request = new HttpPost(baseUri.setPath(endpoint).build)
+        request.addHeader(HttpHeaders.ACCEPT, "application/json")
+        request.setEntity(entity)
+
+        execute(request, timeout.toMillis.toInt)
+    }
+
+    private def execute(request: HttpRequestBase, timeoutMsec: Integer): Either[ContainerConnectionError, ContainerResponse] = {
+        var response: CloseableHttpResponse = null
+        try {
+            response = connection.execute(request)
+            Try(response.getEntity).filter(_ != null).map { entity =>
+                val statusCode = response.getStatusLine.getStatusCode
+                entity.getContentLength match {
+                    case length if length != -1 =>
+                        if (length <= maxResponseBytes) {
+                            val str = EntityUtils.toString(entity, StandardCharsets.UTF_8)
+                            Right(ContainerResponse(statusCode, str))
+                        } else {
+                            // truncate the response
+                            val bytesToRead = Math.min(length, maxResponseBytes)
+                            val bytes = IOUtils.toByteArray(entity.getContent, bytesToRead)
+                            val str = new String(bytes, StandardCharsets.UTF_8)
+                            Right(ContainerResponse(statusCode, str, Some(length.B, maxResponse)))
+                        }
+
+                    case _ => Left(NoResponseReceived())
```

Comment:
This is just the `-1` case? I think this is too weak in general:

```
/**
     * Tells the length of the content, if known.
     *
     * @return  the number of bytes of the content, or
     *          a negative number if unknown. If the content length is known
     *          but exceeds {@link java.lang.Long#MAX_VALUE Long.MAX_VALUE},
     *          a negative number is returned.
     */
```

Why don't we do:
```
val contentLength = entity.getContentLength
if(contentLength > 0 && contentLength <= maxResponseBytes) {
    // good case
} else if(contentLength > maxResponseBytes) {
    // truncated case
} else {
    Left(NoResponseReceived())
}
```

I think that handles the degenerated cases better, as its not specified to return `-1` on every degenerated case? Thoughts?